core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/ForwardChainingSchemaCachingRDFSInferencer.java

```diff
@@ -0,0 +1,318 @@
+            throw new SailException("Connection already has a lock! Might be a begin() without a commit() or rollback() for previous transaction().");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+//        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+        if(connection.lockStamp == 0){
+            throw new SailException("Expected connection to have lock. Might be a commit() without a begin().");
+        }
+        readWriteLock.unlock(connection.lockStamp);
+//        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+
+    private AtomicInteger numberOfThreadsWaitingForWriteLock = new AtomicInteger(0);
+
+    void upgradeLock(ForwardChainingSchemaCachingRDFSInferencerConnection connection) {
+
+//        System.err.println("Attempt writelock: "+connection.lockStamp);
+
+        numberOfThreadsWaitingForWriteLock.incrementAndGet();
+
+        try {
+           while(true) {
+                long l = readWriteLock.tryConvertToWriteLock(connection.lockStamp);
+
+                if (l != 0) {
+//                    long temp = connection.lockStamp;
+                    connection.lockStamp = l;
+//                    if (temp != l) {
+//                        System.err.println("readLock: " + temp + " writeLock: " + connection.lockStamp);
+//                    }
+
+
+                    return;
+                }
+
+                try {
+                    Thread.sleep(random.nextInt(2));
+                } catch (InterruptedException e) {
+                    // ignore interrupted exception
+                }
+
+                // detect potential deadlock scenario
+                if(numberOfThreadsWaitingForWriteLock.get() > 1  //More than 1 connection waiting for a write lock
+                    && numberOfThreadsWaitingForWriteLock.get() <= readWriteLock.getReadLockCount() // no connection only using a read lock
+                    && random.nextBoolean()){ // randomly kick this connection out
+                    break;
+                }
+
+            }
+
+//            System.err.println("isReadLocked: " + readWriteLock.isReadLocked());
+//            System.err.println("isWriteLocked(): " + readWriteLock.isWriteLocked());
+//            System.err.println("read lock count: " + readWriteLock.getReadLockCount());
+
+            releaseLock(connection);
+            throw new SailConflictException("Concurrent modification of schema, could not acquire the lock.");
+        } finally {
+
+            numberOfThreadsWaitingForWriteLock.decrementAndGet();
+        }
+
+    }
+
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data) {
+        super(data);
+        schema = null;
+
+    }
+
+
+    public ForwardChainingSchemaCachingRDFSInferencer(NotifyingSail data, Repository schema) {
+        super(data);
```

Comment:
This should call ```this(data, schema, DEFAULT_USE_RDFS_RULES)```, with a new constant, rather than relying on the initialiser to set it to true, given the constructors set it up in most cases.